lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,393 @@
+  }
+  bool IsLessThanUpperVal(
+      const std::int64_t &intVal, const std::int64_t &upperIntVal) {
+    return intVal <= upperIntVal;
+  }
+  template<typename T>
+  bool CheckOverlappingValue(
+      const T &val, std::optional<T> lowerVal, std::optional<T> upperVal) {
+    return ((!lowerVal || IsMoreThanLowerVal(val, lowerVal.value())) &&
+        (!upperVal || IsLessThanUpperVal(val, upperVal.value())));
+  }
+  template<typename T>
+  bool CheckOverlappingRangeValue(std::optional<T> memberFromVal,
+      std::optional<T> memberToVal, std::optional<T> rangeFromVal,
+      std::optional<T> rangeToVal) {
+    return ((memberFromVal &&
+                CheckOverlappingValue(
+                    memberFromVal.value(), rangeFromVal, rangeToVal)) ||
+        (memberToVal &&
+            CheckOverlappingValue(
+                memberToVal.value(), rangeFromVal, rangeToVal)) ||
+        (rangeFromVal &&
+            CheckOverlappingValue(
+                rangeFromVal.value(), memberFromVal, memberToVal)) ||
+        (rangeToVal &&
+            CheckOverlappingValue(
+                rangeToVal.value(), memberFromVal, memberToVal)));
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::int64_t>::value,
+      std::map<std::int64_t, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseIntVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::string>::value,
+      std::map<std::string, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseStringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u16string>::value,
+      std::map<std::u16string, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseU16StringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::map<std::u32string, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseU32StringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::int64_t>::value,
+      std::vector<SelectCaseRangeType<std::int64_t>> &>::type
+  GetRangeValHandle() {
+    return selectCaseIntRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::string>::value,
+      std::vector<SelectCaseRangeType<std::string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseStringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u16string>::value,
+      std::vector<SelectCaseRangeType<std::u16string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU16StringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::vector<SelectCaseRangeType<std::u32string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU32StringRangeVals_;
+  }
+
+  std::map<bool, parser::CharBlock> selectCaseLogicalVals_;
```

Comment:
`CASE(x)` with a single value and no colon can be represented as the range `x:x`.

It is not necessary for this class to be able to represent maps of all of these types at the same time; all of the cases in a `SELECT CASE` have to hold constant ranges of the same type.  Abstracting the entire class as a template would be clearer.

Reply:
Thanks Klausler, 
>> CASE(x) with a single value and no colon can be represented as the range x:x.
This is a very good suggestion. I will re-implement with this modification.

Reply:
I am stuck while implementing this approach:
I tried to abstract the entire class as a template
template<typename T> class SelectCaseHelper {
  ..
  std::vector<SelectCaseRangeType> selectCaseRangeVals_;
  ..
  }

At the time of creating an object, I need the exact type:
{bool, std::int64_t, std::string, std::u16string, std::u32string}

SelectCaseHelper<Type> selectCaseStmts(context_.messages(),
      selectCaseStmtType->kind(), selectCaseStmtType.value().category());

All we have is the kind and the category. 
Is there a way to get the equivalent type by using the kind and the category?
I tried to figure this out, but couldn't.

Reply:
I will implement it when I get time.

Reply:
I tried to use factory method to create the object with specific types, but this didn't work because of member function templates. So I gave up here and I have added a TODO for this comment.